service/history/nDCActivityReplicator.go

```diff
@@ -184,6 +186,21 @@ func (r *nDCActivityReplicatorImpl) SyncActivity(
 		return err
 	}
 
+	// sync activity with empty started ID means activity retry
+	if request.StartedId == common.EmptyEventID && request.Attempt > activityInfo.GetAttempt() {
```

Comment:
The `Attempt` check needs to happen before `ReplicateActivityInfo` on L164 I believe, as the replicate method will mutate the value pointed to by `activityInfo`